### DISCUSSION TITLE
using storage in functions

### DISCUSSION THREAD

**Discussion by dancixx** [nhost/nhost#2864] [Q&A]
Hi guys,

I am using storage in function to generate a blurhash for an image. If I am using the presignedUrl I am always getting undefined. Any idea why this is happening?
Second, I  switched to publicUrl. Until I didn't provide the accessToken I got 403. Right now I have publicUrl but now I'm getting 500 when I try to download the image using Axios.

Do you have any idea, if this is a good approach to achieve my goal or not?
Thanks for your help.

```ts
import { NhostClient } from "@nhost/nhost-js";
import axios from "axios";
import { encode } from "blurhash";
import { Request, Response } from "express";
import gql from "graphql-tag";
import { Jimp } from "jimp";

export default async (req: Request, res: Response) => {
  console.log("req.body", req.body);
  const nhost = new NhostClient({
    subdomain: process.env.NHOST_SUBDOMAIN,
    region: process.env.NHOST_REGION,
  });

  console.log("env", process.env.NHOST_SUBDOMAIN, process.env.NHOST_REGION);

  try {
    const fileId = req.body.event.data.new.id;
    const url = nhost.storage.getPublicUrl({ fileId });
    const token = nhost.auth.getAccessToken();
    console.log("URL", url);

    const image = await axios.get(url!, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
      responseType: "arraybuffer",
    });
    console.log("image", image);

    const imageBuffer = Buffer.from(image.data);
    const img = await Jimp.fromBuffer(imageBuffer);

    img.resize({ w: 32, h: 32 });

    const pixels: number[] = [];
    img.scan(0, 0, img.bitmap.width, img.bitmap.height, function (x, y, idx) {
      const red = this.bitmap.data[idx];
      const green = this.bitmap.data[idx + 1];
      const blue = this.bitmap.data[idx + 2];
      const alpha = this.bitmap.data[idx + 3];

      pixels.push(red, green, blue, alpha);
    });

    const pixelArray = new Uint8ClampedArray(pixels);
    const blurhash = encode(new Uint8ClampedArray(pixelArray), img.bitmap.width, img.bitmap.height, 4, 4);

    const graphql = gql`
      mutation update_image($id: uuid!, $blurhash: String!) {
        update_profile_images_by_pk(pk_columns: { id: $id }, _set: { blurhash: $blurhash }) {
          id
          blurhash
        }
      }
    `;

    await nhost.graphql.request(graphql, { id: fileId, blurhash });
    res.status(200).json({ message: "Blurhash generated" });
  } catch (error) {
    console.error(error);
    res.status(500).json({ message: "Error generating blurhash" });
  }
};
```

**Comment by dbarrosop** [comment]
> Until I didn't provide the accessToken I got 403

Yes, you need permissions for downloading files

> Right now I have publicUrl but now I'm getting 500 when I try to download the image using Axios

Check the storage logs as there should be more info there. My guess is that your authentication header is incorrect as you are calling `nhost.auth.getAccessToken` but I don't see you signing in anywhere, which means it will be empty.

Assuming the issue is authentication as I suspect, I'd suggest you to look into PATs (personal access token); `createPAT` and `signin({pat})`. You should be able to create the PAT on your dev machine, add it as a secret/environment variable to your functions and sign in with it.

You can also look into `nhost.storage.download` which handles the authorization token automatically (unless you have your reasons to use axios)

**Reply by dancixx** [reply]
When I initialized my client this way, solved all my problems, which is a bit weird for me because I would think that the nhost instance is available out of the box in the functions with full permission, something like in the case of google cloud functions.

```ts
  const nhost = new NhostClient({
    subdomain: process.env.NHOST_SUBDOMAIN,
    region: process.env.NHOST_REGION,
    adminSecret: process.env.NHOST_ADMIN_SECRET,
    storageUrl: process.env.NHOST_STORAGE_URL,
    graphqlUrl: process.env.NHOST_GRAPHQL_URL,
  });
```
